xmtp_mls/src/worker.rs

```diff
@@ -0,0 +1,130 @@
+use crate::{configuration::WORKER_RESTART_DELAY, groups::device_sync::worker::SyncMetric, Client};
+use metrics::WorkerMetrics;
+use parking_lot::Mutex;
+use std::{fmt::Debug, hash::Hash, marker::PhantomData, sync::Arc};
+use xmtp_api::XmtpApi;
+use xmtp_db::XmtpDb;
+
+pub mod metrics;
+
+#[derive(PartialEq, Eq, Hash)]
+pub enum WorkerKind {
+    DeviceSync,
+    DisappearingMessages,
+    KeyPackageCleaner,
+}
+
+pub(crate) trait WorkerManager: Send + Sync {
+    fn sync_metrics(&self) -> Option<Arc<WorkerMetrics<SyncMetric>>>;
+    fn spawn(&self);
+}
+
+impl<Core, Metric> WorkerManager for WorkerRunner<Core, Metric>
+where
+    Core: Send + Sync,
+    Metric: Send + Sync + 'static,
+{
+    fn sync_metrics(&self) -> Option<Arc<WorkerMetrics<SyncMetric>>> {
+        if std::any::TypeId::of::<Metric>() == std::any::TypeId::of::<SyncMetric>() {
+            self.metrics.as_ref().lock().clone().map(|arc| {
+                // This is safe because we verified Metric == SyncMetric
+                unsafe { std::mem::transmute::<Arc<WorkerMetrics<Metric>>, Arc<WorkerMetrics<SyncMetric>>>(arc.clone()) }
```

Comment:
This feels like a shortcut/smell that shouldn't be necessary/anti-pattern. Why can't we define what a metric should return and enforce that, rather than do an unsafe cast to a concrete type?

Or, just enforce a concrete type on the trait, rather than an opaque type. Could we come up with a metric type generic enough to work for all workers? What metrics are useful here

I already approved so its too late for this PR, but we should aim to avoid unsafe, even if it looks OK on paper. This could fail down the road in ways we don't expect, and lead to some very difficult to debug issues